cpp/src/parquet/statistics.cc

```diff
@@ -567,6 +568,27 @@ class TypedStatisticsImpl : public TypedStatistics<DType> {
     SetMinMaxPair(comparator_->GetMinMax(values));
   }
 
+  void UpdateArrowDictionary(const ::arrow::Array& indices,
+                             const ::arrow::Array& dictionary) {
+    IncrementNullCount(indices.null_count());
+    IncrementNumValues(indices.length() - indices.null_count());
+
+    if (indices.null_count() == indices.length()) {
+      return;
+    }
+
+    ::arrow::compute::ExecContext ctx(pool_);
+    PARQUET_ASSIGN_OR_THROW(auto referenced_indices,
+                            ::arrow::compute::Unique(indices, &ctx));
+    PARQUET_ASSIGN_OR_THROW(
```

Comment:
this allocate a whole new array?  maybe we should file a follow-up JIRA to make this more efficient.  

In particular it seems like it could be more efficient to get map of dictionary index to sort order (I think we already have a kernel for this) if necessary each time before calculating the statistics if there are any new entries then iterate through the indices doing a comparison in index space.

Reply:
I'll create a follow-up.  Another approach I considered was that we could compute the unique indices and then pass those down as a selection filter to the GetMinMax function.  My goal was minimizing change to the existing code.

Or, since we have all these compute kernels now, just use ::arrow::compute::MinMax and pass the min & max into the statistics directly.

Reply:
If we have MinMax and it works for dictionary arrays that sounds like a good idea to me.

Reply:
First, I'm not sure if it was clear, but there are actually two array allocations here.  First it creates a unique array which is a (hopefully small) subset of the indices.  Second, it creates a referenced values array which is a (probably not much smaller) subset of the values.  Sadly the MinMax functions don't yet support strings.  There are kernels (I think) to get a sort order but I can't see how that will help because they don't take selection vectors yet.  We could take the output from `Unique` improve Comparator to take a selection vector.  This would save the second allocation.  However, I'm reluctant to undertake that work when I think the eventual solution will be for the MinMax kernels to support all types.  I've created PARQUET-2068 for follow-up.